src/realm/group_shared.cpp

```diff
@@ -1197,6 +1198,11 @@ void SharedGroup::do_open(const std::string& path, bool no_create_file, bool is_
     // Upgrade file format and/or history schema
     try {
         using gf = _impl::GroupFriend;
+        if (stored_hist_schema_version == -1) {
+            // current_hist_schema_version has not been read. Read it now
+            ReadTransaction rt(*this);
+            stored_hist_schema_version = gf::get_history_schema_version(m_group.m_alloc, m_read_lock.m_top_ref);
+        }
```

Comment:
It really bugs me that an extra read transaction is required here (the cost of it). My gut feeling is that it can be avoided by rearranging some stings, but I cannot say how at this point. One idea, though, is that if we integrate the upgrade process into the session initiation process, then the need for an extra read transaction would go away.